### PR TITLE
CSH-8382: roll back 523b8ba and solve in a different way

### PR DESCRIPTION
**Pull request by wwjhu**
Fixes the problem in ecs backend where the filePaths and getFileContent paths no longer match because one is normalized and the other isn't. Also keeps the getFileContent, getFileSize and filePaths arguments of the validate function consistent: all take normalized paths. The problem of having a path.normalize in the browser is solved by separating the loading of the renditions is a separate function that is not imported in the browser when using parseDefinition. Additionally adds deepFreeze to ensure the components definition is not altered while parsing it into a component set.

### CODE REVIEW

File: lib/models/get-file-content.ts [hunk 1]
```diff
@@ -6,9 +6,12 @@ export type GetFileContentOptionsType = { encoding: string | null } | null;
 /**
  * Return string if encoding is string and Buffer otherwise.
  */
-export type GetFileContentType = (filePath: string, options?: GetFileContentOptionsType) => Promise<string | Buffer>;
+export type GetFileContentType = (
+    normalizedRelativeFilePath: string,
```
Comment (ruudjungbacker) [inline lib/models/get-file-content.ts:10]: I like the changes of these parameter names. But should they also be updated in the readme? The readme only mentions:
`const getFileContent = async (filePath: string) => 'fileContent';`

It doesn't mention the options at all (which isn't that important I think). but the name could be updated to normalizedRelativeFilePath maybe?

File: lib/renditions/utils.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import * as path from 'path';
+import { Component, ComponentRendition, ComponentsDefinition, GetFileContentType } from '../models';
+import merge = require('lodash.merge');
+import { deepFreeze } from '../util/freeze';
+
+export async function loadRenditions(
```
Comment (ruudjungbacker) [inline lib/renditions/utils.ts:6]: Not sure, but shouldn't this be `loadHtmlRenditions`? as the PSV one is left out of scope.

File: lib/renditions/utils.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+async function loadRendition(
+    component: Component,
+    rendition: ComponentRendition,
+    getFileContent: GetFileContentType,
+): Promise<Component> {
+    if (!component.renditions) {
+        component.renditions = {};
+    }
+    component.renditions[rendition] = (await getFileContent(
+        path.normalize(`./templates/${rendition}/${component.name}.html`),
```
Comment (ruudjungbacker) [inline lib/renditions/utils.ts:32]:
```suggestion
        path.normalize(`templates/${rendition}/${component.name}.html`),
```

File: lib/renditions/utils.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import * as path from 'path';
+import { Component, ComponentRendition, ComponentsDefinition, GetFileContentType } from '../models';
+import merge = require('lodash.merge');
+import { deepFreeze } from '../util/freeze';
+
+export async function loadRenditions(
+    componentsDefinition: ComponentsDefinition,
+    getFileContent: GetFileContentType,
+): Promise<ComponentsDefinition> {
+    const enrichedDefinition = merge({}, componentsDefinition);
```
Comment (ruudjungbacker) [inline lib/renditions/utils.ts:10]: Just for my understanding, this 'unfreezes' an object?

**Review by ruudjungbacker — COMMENTED** [review]
I added a few comments/questions, but I like the overal solution for the issue!

File: lib/renditions/utils.ts (hunk 2 above)
Comment (wwjhu) [reply lib/renditions/utils.ts:6]: Makes sense.

File: lib/renditions/utils.ts (hunk 4 above)
Comment (wwjhu) [reply lib/renditions/utils.ts:10]: It creates a [copy of the object](https://lodash.com/docs#merge)

**Review by ruudjungbacker — APPROVED** [review]
🚀

File: lib/renditions/utils.ts (hunk 4 above)
Comment (wwsno) [reply lib/renditions/utils.ts:10]: Shouldn't you use lodash deepClone in this case? Or is it different?

**Review by wwsno — APPROVED** [review]
Looks fine 👍

File: lib/renditions/utils.ts (hunk 4 above)
Comment (wwjhu) [reply lib/renditions/utils.ts]: The result is the same. I copied this from the old code, but I guess that for new code we're better off using the function that describes best what it does: cloneDeep.

**Pull request merged**